pkgs/stdenv/generic/make-derivation.nix

```diff
@@ -259,13 +266,16 @@ else let
     lib.unique (lib.concatMap (input: input.__propagatedImpureHostDeps or [])
       (lib.concatLists propagatedDependencies));
 
+  envIsExportable = lib.isAttrs env && !lib.isDerivation env;
+
   derivationArg =
     (removeAttrs attrs
-      ["meta" "passthru" "pos"
+      (["meta" "passthru" "pos"
        "checkInputs" "installCheckInputs"
        "__darwinAllowLocalNetworking"
        "__impureHostDeps" "__propagatedImpureHostDeps"
-       "sandboxProfile" "propagatedSandboxProfile"])
+       "sandboxProfile" "propagatedSandboxProfile"]
```

Comment:
Now that we have `env`, shouldn't we use `attrs` to define environment variables?
This out-of-place list of variables is another symptom of the multi-purposing problem. See https://github.com/NixOS/nixpkgs/pull/175649/files#r909504287

Reply:
I suppose we could have three levels of `structuredAttrs`:
 - `structuredAttrs = false;`: status quo
 - `structuredAttrs = "lax";`: set `__structuredAttrs = true;` and cause all `attrs` to be `derivation` args according to the current expression
 - `structuredAttrs = true;`: set `__structuredAttrs = true;` and remove all effects of `attrs` attributes that aren't known `mkDerivation` parameters, making them local variables that are only accessible to `overrideAttrs`, which is currently a missing feature. I.e, conditionally change this from `removeAttrs attrs x // y` to effectively `y`.

This change will allow `mkPythonPackage` and similar functions to be defined in terms of `overrideAttrs` when the migration to `structuredAttrs` is complete, simplifying their code and removing the need for language-specific `overrideAttrs`-like functions.

Reply:
(I've tried to change `mkPythonPackage` to `pythonBasePackage.overrideAttrs`, but got blocked on the inability to control which arguments become an actual part of the derivation and how. I don't have proof that my proposal is sufficient, but it must be close.)

Reply:
> Now that we have `env`, shouldn't we use `attrs` to define environment variables? This out-of-place list of variables is another symptom of the multi-purposing problem. See [#175649 (files)](https://github.com/NixOS/nixpkgs/pull/175649/files#r909504287)

I'm not sure what you mean?

Reply:
Oh, the link is broken now :(.
I meant to say that since we have `env`, we can stop using `attrs` to define environment variables, and we probably should. 
If we make such a change, the `mkDerivation` arguments become a record instead of a record+dictionary, which is a significant quality improvement.

Reply:
I'm resolving this conversation because of scope creep.

Structured attrs is primarily a builder-level change. My suggestion here to change `mkDerivation` is both too ambitious for bundling in with this PR and not ambitious enough by itself. It is better to do it later, perhaps bundled with more improvements.